message/history/eventFinder.go

```diff
@@ -0,0 +1,182 @@
+		return checkEvent(latestEvent, currentBootTime)
+	}
+}
+
+// SameEventFinder returns a function that goes through a list of events and compares the currentEvent
+// to these events to make sure that currentEvent is valid. If any of the fatalValidators returns false,
+// it will stop searching and immediately exit, returning the event
+// that prompted the error along with the error returned by the fatalValidator. If all of the fatalValidators
+// pass, the currentEvent is returned along with nil error.
+func SameEventFinder(fatalValidators validation.Validator) FinderFunc {
+	return func(events []message.Event, currentEvent message.Event) (message.Event, error) {
+		// verify that the current event has a boot-time
+		currentBootTime, err := currentEvent.BootTime()
+		if currentBootTime <= 0 {
+			return message.Event{}, validation.InvalidBootTimeErr{OriginalErr: err}
+		}
+
+		for _, event := range events {
+			// if transaction UUIDs are the same, continue onto next event
+			if event.TransactionUUID == currentEvent.TransactionUUID {
+				continue
+			}
+			// if any fatalValidators return false, it means we should stop looking for an event
+			// because there is something wrong with currentEvent, and we should not
+			// perform calculations using it.
+			if valid, err := fatalValidators.Valid(event); !valid {
+				return event, validation.InvalidEventErr{OriginalErr: err}
+			}
+		}
+
+		return currentEvent, nil
+	}
+}
+
+// See if event has a boot-time that has greater than the one we are currently tracking but less than
+// the latestBootTime.
+func getPreviousBootTime(event message.Event, currentPrevTime int64, latestBootTime int64) int64 {
+	// Get the bootTime from the event we are checking. If boot-time
+	// doesn't exist, return currentPrevTime, which is the latest previous time currently found.
+	bootTime, _ := event.BootTime()
+	if bootTime <= 0 {
+		return currentPrevTime
+	}
+
+	// if boot-time is greater than any we've found so far but less than the current boot-time,
+	// return bootTime
+	if bootTime > currentPrevTime && bootTime < latestBootTime {
+		return bootTime
+	}
+	return currentPrevTime
+}
+
+// Sees if an event is valid based on the validators passed in and whether it has the targetBootTime.
+// If not, returns defaultEvent.
+func compareEvents(newEvent message.Event, defaultEvent message.Event, validators validation.Validator, targetBootTime int64) message.Event {
+	bootTime, _ := newEvent.BootTime()
+	currentPrevBootTime, _ := defaultEvent.BootTime()
+
+	// if boot-time doesn't match target boot-time, return previous event
+	if bootTime != targetBootTime {
+		return defaultEvent
+	}
+
+	// if event does not match validators, return previous event
+	if valid, _ := validators.Valid(newEvent); !valid {
+		return defaultEvent
+	}
+
+	if currentPrevBootTime != targetBootTime || newEvent.Birthdate < defaultEvent.Birthdate {
+		return newEvent
+	}
+
+	return defaultEvent
+}
```

Comment:
I think if this function returned a boolean, transferring the responsibility of what event should be sent to the caller's variable, the caller would be able to also take other actions depending on the result of this function call.  This would help us, because then the caller knows when `latestEvent` has been set, removing the need for `checkEvent`.

Reply:
The changes needed for this function would be change the name (to something like `EarlierValid(....) bool`) and then return `true` instead of `newEvent`, and `false` instead of `defaultEvent`.

Then, in `CurrentSessionFinder()`, the code could do something like:
```
func CurrentSessionFinder(...) ... {
  var latestEvent message.Event
  var found
  for _, e := range events {
    // ...
    if earlier := EarlierValid(e, ....); earlier {
      latestEvent = e
      found = true
    }
  }
  if !found {
    return EventNotFoundErr
  }
  return nil
}
```

Reply:
For `LastSessionFinder()`:
```
func LastSessionFinder( ... ) ... {
  var latestEvent message.Event
  var prevBootTime int64
  var found bool
  for _, e := range events {
    // ...
    // this requires changing getPreviousBootTime() to take e's boottime and return a boolean
    eBoot := e.BootTime()
    if later := laterBootTime(eBoot, prevBootTime, currentBootTime); later {
      prevBootTime = eBoot
      found = false
    }
    if earlier := earlierValid(e, .....); earlier {
      latestEvent = e
      found = true
    }
  }
  if !found {
    return EventNotFoundErr
  }
  return nil
}
```